### PR TITLE
Implemented Vector Class

### PR DESCRIPTION
**Pull request by LegenDAirie**
This is my first attempt at making a vector class that mimics the built-in vector class in processing. Please review.

### CODE REVIEW

File: ball.py [hunk 1]
```diff
@@ -1,10 +1,76 @@
 from random import randint, random
+from vector import Vector
 from sketchy import Sketchy
 import math
 
 class Ball(object):
+    def __init__(self, width, height):
+         self.location = Vector(width/2, height/2)
+         self.velocity = Vector(0, 0)
+        #  self.center = Vector(width/2, height/2)
```
Comment (rsandor) [inline ball.py]: Do you need to leave these comments all over the place? Have you considered setting up a specific test directory that just implements your vector tests?

File: main.py [hunk 2]
```diff
@@ -1,15 +1,20 @@
 from sketchy import Sketchy
 from ball import Ball
+from vector import Vector
 
 class MyDrawing(Sketchy):
 
     def setup(self):
-        return
+        width, height = [600, 400]
```
Comment (rsandor) [inline main.py]: Are you planning on using `ball.py` and `main.py` to test every new feature you make? If so consider having different implementations for each test so you have the tests later when you are making additions or fixes to `vector.py`.

File: vector.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    # mul(scalar)
+    def mul(self, scalar):
+        return Vector(self.x * scalar, self.y * scalar)
+
+    # a * scalar
+    def __mul__(self, scalar):
+        return self.mul(scalar)
+
+    # a *= scalar
+    def multAccumulate(self, scalar):
```
Comment (rsandor) [inline vector.py]: Shouldn't this be named `mulAccumulate`? All of the multiplication methods in the rest of the class use `mul` as a prefix and not `mult`.

File: vector.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        self.x *= x
+        self.y *= y
+        return self
+
+    # a *= scalar
+    def __imul__(self, scalar):
+        return self.multAccumulate(scalar)
```
Comment (rsandor) [inline vector.py]: If you rename to `mulAccumulate`, don't forget to change it here.

File: vector.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        self.x /= scalar
+        self.y /= scalar
+        return self
+
+    # a /= scalar
+    def __idiv__(self, scalar):
+        return self.divAccumulate(scalar)
+
+    # get_mag()
+    def getMag(self):
```
Comment (rsandor) [inline vector.py]: Perhaps this would be better if it was called `magnitude`? I don't like it being called `getMag` because none of the other methods use the prefix `get`.

File: vector.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        hyp = math.sqrt((self.x ** 2) + (self.y ** 2))
+        return hyp
+
+    # norm()
+    def norm(self):
```
Comment (rsandor) [inline vector.py]: Perhaps you could be more specific with this name and call it `normalize`, what do you think?

File: vector.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        hyp = self.getMag()
```
Comment (rsandor) [inline vector.py]: Why did you name this variable `hyp`, out of curiosity?

File: vector.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        if hyp != 0:
+            self.x /= hyp
+            self.y /= hyp
+            return Vector(self.x, self.y)
+
+    def setMag(self, value):
```
Comment (rsandor) [inline vector.py]: Doesn't this seem a bit redundant? Consider this:

``` python
# Create a vector...
v = Vector(10, 0)

# Next two lines do the same thing
v.setMag(5)
v = v.norm() * 5
```

(assuming your `.norm` method returns a new vector instead of setting the vector's properties).

I am of the opinion that the second form is a little better because it explicitly shows what is going on without an application programmer having to know anything about `setMag`. Thoughts?

File: vector.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        mag = self.getMag()
+        self.x *= value / mag
+        self.y *= value / mag
+        return Vector(self.x, self.y)
+
+    def limit(self, topSpeed):
```
Comment (rsandor) [inline vector.py]: Doesn't this seem a bit too specific for a general vector math class? If I recall correctly, you were using this to limit the speed for a velocity vector. In this case it is a very specific operation and doesn't have much to do with general vectors (as it is implementation dependent). 

Perhaps you could create a utility library that includes a limit function for limiting vectors in the future?

File: vector.py (hunk 6 above)
Comment (rsandor) [inline vector.py]: Also I don't think it is clear by the name that you're actually changing the state of the vector here. I would assume that `.norm` or `.normalize` would just return the normalized version of the vector and not change its actual magnitude. 

Perhaps this should be two functions? One that returns the normal for a vector, and another that sets the normal? On the other hand you can already assign vectors, so perhaps just make this method return a new vector that is normal, and then you could use the following to actually set it to the vector:

``` python
v = v.norm()
```

**Comment by rsandor** [comment]
This is really solid work, good job! There are a few design issues, however, and I have mentioned them on specific lines during my review.

**Comment by rsandor** [comment]
Additionally there is no documentation (in the form of `"""` block comments as the first statement in each function). I'd suggest that you document each of the methods before we merge this into master.

File: ball.py (hunk 1 above)
Comment (LegenDAirie) [inline ball.py]: I don't know what you mean by setting up a specific test directory and implementing my vector tests.

File: main.py (hunk 2 above)
Comment (LegenDAirie) [inline main.py]: I am more so using ball.py and main.py to play around and see what i can do using the vector library.

File: vector.py (hunk 6 above)
Comment (LegenDAirie) [inline vector.py]: Are you suggesting that it returns the result but does not automatically change it to the result? and if so i am not quite sure i know how to do that.

File: vector.py (hunk 7 above)
Comment (LegenDAirie) [inline vector.py]: hypotenuse

File: vector.py (hunk 8 above)
Comment (LegenDAirie) [inline vector.py]: multiplying a normalized vector and setting its magnitude seem to have different effects for me since i use normalize more so for direction and add that normalized vector to a second vector and set the magnitude of the second vector so that magnitude doesn't get to high.

File: vector.py (hunk 9 above)
Comment (LegenDAirie) [inline vector.py]: Are you talking about separating this into a general vector math library and another file that utilizes them in a specific way?

**Comment by LegenDAirie** [comment]
Is it considered good practice to use """ block comments in every function describing what it does and/or how it does it?

File: main.py (hunk 2 above)
Comment (rsandor) [inline main.py]: Do you want the tests to persist though?

File: vector.py (hunk 9 above)
Comment (rsandor) [inline vector.py]: I'm saying it doesn't seem to fit the design. Limiting is a corner-case operation for the physics you are trying to emulate, not an inherent behavior of all mathematical vectors. Consider removing this method for now and if you really need it, creating a Physics utility library that includes it as a class method or something.

**Comment by rsandor** [comment]
@sundoresu - Docstring block comments are a [standard in python](https://www.python.org/dev/peps/pep-0257/). For example, you would document methods and classes like so:

``` python
class Widget(object):
  """ Basic widget class used for all widget types in the project """

  def combine(self, w):
    """ Combines this widget with another widget and returns a new widget. """

  def disassemble(self):
    """
    This method meticulously disassembles the widget. Disassembly can be performed
    all at once or in parts depending on the nature of the specific widget. Subclasses 
    should override this method to perform the desired behavior.
    """
```

The first `"""` denotes the documentation string for the `Widget` class itself. The second for the `combine` method. The third is a multi-line docstring, notice how the `"""` each appear on their own separate lines.

File: main.py (hunk 2 above)
Comment (LegenDAirie) [inline main.py]: Do you mean make tests? since there is currently none.

File: vector.py (hunk 9 above)
Comment (LegenDAirie) [inline vector.py]: Ok, thats a good idea.

**Comment by LegenDAirie** [comment]
I abstracted out limit into its own function in the file for physics simulation stuff, and the program runs and it looks how it is suppose to but i get an error that i dont understand. if you run it, it reads
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 314, in 'calling callback function'
  File "/Users/airiel/Projects/The_Nature_of_Code/sketchy/vector_test/sketchy.py", line 153, in _render
    self.draw(self.g)
  File "main.py", line 19, in draw
    self.ball.draw(g)
  File "/Users/airiel/Projects/The_Nature_of_Code/sketchy/vector_test/ball.py", line 31, in draw
    g.circle(self.mouse.x, self.mouse.y, 20)
AttributeError: 'Ball' object has no attribute 'mouse'

File: vector.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
+from random import randint, random
+import math
+
+class Vector(object):
```
Comment (rsandor) [inline vector.py]: Consider renaming the class to Vector2D

File: vector.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        return hypotenuse
+
+    # normalize()
+    def normalize(self):
+        """ reduces the vector to unit vector form, a magnitude of 1 """
+        hypotenuse = self.magnitude()
+        if hypotenuse != 0:
+            self.x /= hypotenuse
+            self.y /= hypotenuse
+            return Vector(self.x, self.y)
```
Comment (rsandor) [inline vector.py]: This is still setting the `self.x` and `self.y` of the object. I think you meant:

``` python
if hypotenuse != 0:
  return Vector(self.x / hypotenuse, self.y / hypotenuse)
```

Comment (rsandor) [inline vector.py]: Also this should return `Vector(0, 0)` if the hypotenuse is zero.

File: main.py (hunk 2 above)
Comment (rsandor) [inline main.py]: Don't worry about this for now, I will show you how to setup unit tests for the class later.

File: lib/vector.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    def magnitude(self):
+        """ tells you the magnitude of a vector """
+        hypotenuse = math.sqrt((self.x ** 2) + (self.y ** 2))
+        return hypotenuse
+
+    # normalize()
+    def normalize(self):
+        """ reduces the vector to unit vector form, a magnitude of 1 """
+        hypotenuse = self.magnitude()
+        if hypotenuse != 0:
```
Comment (rsandor) [inline lib/vector.py]: Change this to:
```python
if hypotenuse != 0:
  return Vector(self.x / hypotenuse, self.y / hypotenuse)
else:
  return Vector(0, 0)
```

**Comment by rsandor** [comment]
LGTM :+1: :100:
